project_recalculate/models/project_task.py

```diff
@@ -0,0 +1,168 @@
+                date += timedelta(days=1)
+            else:
+                date -= timedelta(days=1)
+        return date
+
+    def _calculate_date_without_weekend(self, date_start, days,
+                                        increment=True):
+        total_days = 0
+        c = 0
+        if days < 0:
+            if increment:
+                increment = False
+            else:
+                increment = True
+        days = days if days >= 0 else days * -1
+        first = True
+        while days != total_days:
+            if total_days > 0 or not first:
+                recalculate = days + (days - total_days) + c
+                c += 1
+            else:
+                first = False
+                recalculate = days
+            if increment:
+                start = date_start
+                end = date_start + timedelta(days=recalculate)
+            else:
+                start = date_start - timedelta(days=recalculate)
+                end = date_start
+            total_days = self._count_days_without_weekend(start, end)
+        if total_days > 0:
+            date = end if increment else start
+        else:
+            date = date_start
+        return self._correct_days_to_workable(date, increment)
+
+    def on_change_dates(self, date_start, date_end, vals):
+        if not date_start or not date_end:
+            return vals
+        start = fields.Date.from_string(date_start)
+        end = fields.Date.from_string(date_end)
+        end += timedelta(1)
+        vals['estimated_days'] = self._count_days_without_weekend(start, end)
+        calculation_type = self.project_id.calculation_type
+        if calculation_type:
+            if calculation_type == 'date_begin':
+                if not self.project_id.date_start:
+                    return vals
+                aux_start = fields.Date.from_string(self.project_id.date_start)
+                aux_end = start
+            else:
+                if not self.project_id.date:
+                    return vals
+                aux_start = start
+                aux_end = fields.Date.from_string(self.project_id.date)
+            vals['from_days'] = self._count_days_without_weekend(
+                aux_start, aux_end)
+        return vals
+
+    @api.multi
+    def write(self, vals):
+        if (not self.env.context.get('project_recalculate')
+                and (vals.get('date_start') or vals.get('date_end'))):
+            date_start = (vals.get('date_start')
+                          if vals.get('date_start') else self.date_start)
+            date_end = (vals.get('date_end')
+                        if vals.get('date_end') else self.date_end)
+            vals = self.on_change_dates(date_start, date_end, vals)
+        if 'estimated_days' in vals and vals['estimated_days'] < 1:
+            vals['estimated_days'] = 1
+        return super(ProjectTask, self).write(vals)
+
+    def _tz_date_recalculate(self, date):
+        tz = self.env.user.tz
```

Comment:
At first instance, I should say that the recalculation should be timezone agnostic, because you make all the changes on UTC.